### PR TITLE
 [Cards] accessibility example for collection cards

### PR DESCRIPTION
**Pull request by galiak11**
Updating the "Edit/Recorder" Collection example to demonstrate accessibility for tappable cards with content in a collection (adding an image and a label to cards in a collection). In continuation to documentation in PR: [`[Cards] Add accessibility documentation`](https://github.com/material-components/material-components-ios/pull/4454/files).

image

### CODE REVIEW

File: components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift [hunk 1]
```diff
@@ -0,0 +1,93 @@
+//
+//  CardEditReorderCollectionCell.swift
+//  CatalogByConvention
+//
+//  Created by Galia Kaufman on 7/6/18.
+//
```
Comment (codeman7) [inline components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift]: We need to fix the copyright

**Comment by romoore** [comment]
Thanks, @galiak11 . Could you please post a screenshot or two of the changes?

**Comment by romoore** [comment]
Where did you get the images from? Can you please document the source and make sure that we have the correct license/ownership to use these images?

**Comment by galiak11** [comment]
Please see attached screen shot.
The images are taken from the [Material Design Guidelines page](https://material.io/design/components/cards.html#card-collections). 
I've also added this link to the source file.

File: components/Cards/examples/resources/CardsAssets.xcassets/cell-image-3.imageset/Contents.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{
+  "images" : [
+    {
+      "idiom" : "universal",
+      "scale" : "1x"
+    },
+    {
+      "idiom" : "universal",
+      "filename" : "cell-image-3.png",
```
Comment (romoore) [inline components/Cards/examples/resources/CardsAssets.xcassets/cell-image-3.imageset/Contents.json]: Are the 1x and 3x images missing?

File: components/Cards/examples/resources/CardsAssets.xcassets/check-circle-blue.imageset/Contents.json [hunk 3]
```diff
@@ -0,0 +1,22 @@
+{
```
Comment (romoore) [inline components/Cards/examples/resources/CardsAssets.xcassets/check-circle-blue.imageset/Contents.json]: Can we use [Material Icons](https://material.io/tools/icons/) instead? I believe for the catalog, we try to work with as many of the baseline styles/components as possible.  

If we need to have a circular background for the checkmark, can we look at adding a CALayer or something to the UIImageView instead?

File: components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+//  CardEditReorderCollectionCell.swift
+//  CatalogByConvention
+//
+//  Created by Galia Kaufman on 7/6/18.
```
Comment (romoore) [inline components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift]: Please fix the copyright notice.

File: components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+//
+//  Card images source:
+//  Material Design Guidelines (https://material.io/design/components/cards.html#card-collections)
```
Comment (romoore) [inline components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift]: I can't seem to find the asset downloads on that site. Can we chat offline so I can understand how to get the files?

File: components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    let label = UILabel(frame: CGRect.zero)
+    label.numberOfLines = 1
+    label.textAlignment = .center
+    label.lineBreakMode = .byTruncatingTail
+    label.font = self.typographyScheme.caption
+    label.translatesAutoresizingMaskIntoConstraints = false
+    return label
+  }()
+
+  func configure(title: String, number: Int) {
```
Comment (romoore) [inline components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift]: I think it would be better for the CollectionViewController to set the image property instead. It seems a little strange for the view to know how to select its image.

File: components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    } else {
+      pre_iOS11Constraints()
+    }
+  }
+
+  @available(iOS 11, *)
+  private func post_iOS11Constraints() {
+    #if swift(>=3.2)
+      let guide = self.safeAreaLayoutGuide
+      let margin: CGFloat = 8.0
```
Comment (romoore) [inline components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift]: Maybe make these margins into constants used across both functions?

File: components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+      NSLayoutConstraint.activate([
+        imageView.topAnchor.constraint(equalTo: guide.topAnchor, constant: margin),
+        imageView.leadingAnchor.constraint(equalTo: guide.leadingAnchor, constant: margin),
+        imageView.trailingAnchor.constraint(equalTo: guide.trailingAnchor, constant: -margin),
+
+        imageView.bottomAnchor.constraint(equalTo: titleLabel.topAnchor, constant: -thinMargin),
+
+        titleLabel.leadingAnchor.constraint(equalTo: guide.leadingAnchor, constant: margin),
+        titleLabel.trailingAnchor.constraint(equalTo: guide.trailingAnchor, constant: -margin),
```
Comment (romoore) [inline components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift]: Should the constants be adjusted for RTL?

File: components/Cards/examples/EditReorderCollectionViewController.swift [hunk 9]
```diff
@@ -60,8 +60,13 @@ class EditReorderCollectionViewController: UIViewController,
       collectionView.addGestureRecognizer(longPressGesture)
     }
 
+    let titles = ["Mainsail", "Break point", "Matchday", "Blue run", "90th Minute",
+                  "Race track", "Backcourt", "Alley-oop", "Approach", "Scoring",
+                  "Master point", "Overhead", "Relay", "Tie breaker", "Break shot",
+                  "Track race", "Wide ball", "Sideback", "Baseline", "Off side"]
+
     for i in 0..<20 {
```
Comment (romoore) [inline components/Cards/examples/EditReorderCollectionViewController.swift]: Maybe use `titles.count` instead of hard-coding 20?

File: components/Cards/examples/EditReorderCollectionViewController.swift [hunk 10]
```diff
@@ -27,7 +27,7 @@ class EditReorderCollectionViewController: UIViewController,
 
   let collectionView = UICollectionView(frame: .zero,
                                         collectionViewLayout: UICollectionViewFlowLayout())
-  var dataSource = [(Int, Bool)]()
+  var dataSource = [(Int, Bool, String)]()
```
Comment (romoore) [inline components/Cards/examples/EditReorderCollectionViewController.swift]: Looks like the `Bool` isn't used. Can we remove it?

**Review by romoore — DISMISSED** [review]

File: components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift (hunk 5 above)
Comment (galiak11) [reply components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift]: Edit(rsmoore): We're figuring-out the licensing for the imagery.

File: components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift (hunk 6 above)
Comment (galiak11) [reply components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift]: Makes sense.

File: components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift (hunk 8 above)
Comment (galiak11) [reply components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift]: Are you sure that is needed?  I believe RTL uses the same rules in reverse, so it should be use same numbers. Anyway, I tested the layout with RTL setting and the layout seemed to work fine.

File: components/Cards/examples/EditReorderCollectionViewController.swift (hunk 9 above)
Comment (galiak11) [reply components/Cards/examples/EditReorderCollectionViewController.swift]: sure

File: components/Cards/examples/EditReorderCollectionViewController.swift (hunk 10 above)
Comment (galiak11) [reply components/Cards/examples/EditReorderCollectionViewController.swift]: It is now used in new PR.

File: components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift (hunk 7 above)
Comment (galiak11) [reply components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift]: It will be easier to remove the pre-ios11 code when we drop support - if the code is completely separate.

Comment (ianegordon) [reply components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift]: As this is sample code, I'm not inclined to block landing the PR on this.

**Review by ianegordon — APPROVED** [review]

File: components/Cards/examples/EditReorderCollectionWithImagesViewController.swift [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+import UIKit
+
+class EditReorderCollectionWithImagesViewController: UIViewController,
+  UICollectionViewDelegate,
+  UICollectionViewDataSource,
+UICollectionViewDelegateFlowLayout {
```
Comment (romoore) [inline components/Cards/examples/EditReorderCollectionWithImagesViewController.swift]: Nit/Indentation. Should all of these continued lines be 4 spaces from the starting line?

File: components/Cards/examples/EditReorderCollectionWithImagesViewController.swift [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+
+  enum ToggleMode: Int {
```
Comment (romoore) [inline components/Cards/examples/EditReorderCollectionWithImagesViewController.swift]: Why an `enum` with only one value, can we use a Bool instead?

**Review by romoore — CHANGES_REQUESTED** [review]
Let's please wait until we know what the licensing for the photography will be.

File: components/Cards/examples/EditReorderCollectionWithImagesViewController.swift [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+    ("bangkok-chatuchak-market", "Chatuchak Market", false),
+    ("austin-old-west-austin", "Old West Austin", false),
+    ("austin-dawson", "Dawson", false),
+    ("bangkok-aree-flower", "Bangkok Aree", false),
+    ("austin-north-loop", "North Loop", false),
+    ("bangkok-aree-market", "Bangkok Aree", false)
+  ]
+  var longPressGesture: UILongPressGestureRecognizer!
+  var toggle = ToggleMode.reorder
+  var toggleButton: UIButton!
```
Comment (romoore) [inline components/Cards/examples/EditReorderCollectionWithImagesViewController.swift]: Can we use an MDCButton for this?

File: components/Cards/examples/EditReorderCollectionWithImagesViewController.swift [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+  var colorScheme = MDCSemanticColorScheme()
+  let cardScheme = MDCCardScheme();
```
Comment (romoore) [inline components/Cards/examples/EditReorderCollectionWithImagesViewController.swift]: Please add a `var typographyScheme` so it can be set by the catalog.

File: components/Cards/examples/EditReorderCollectionWithImagesViewController.swift [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+
+  override func viewDidLoad() {
+    super.viewDidLoad()
+
+    cardScheme.colorScheme = colorScheme
+    collectionView.frame = view.bounds
+    collectionView.dataSource = self
+    collectionView.delegate = self
+    collectionView.backgroundColor = UIColor(white: 0.9, alpha: 1)
```
Comment (romoore) [inline components/Cards/examples/EditReorderCollectionWithImagesViewController.swift]: Please use `colorScheme.backgroundColor`.

File: components/Cards/examples/EditReorderCollectionWithImagesViewController.swift [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+    collectionView.translatesAutoresizingMaskIntoConstraints = false
+    collectionView.allowsMultipleSelection = true
+    view.addSubview(collectionView)
+
+    navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Reorder",
+                                                        style: .plain,
+                                                        target: self,
+                                                        action: #selector(toggleModes))
+
+    if #available(iOS 9.0, *) {
```
Comment (romoore) [inline components/Cards/examples/EditReorderCollectionWithImagesViewController.swift]: What happens on iOS 8? Is long-press automatically handled?

File: components/Cards/examples/EditReorderCollectionWithImagesViewController.swift [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+                                                            metrics: nil,
+                                                            views: ["view": collectionView]));
+    self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|[view]|",
+                                                            options: [],
+                                                            metrics: nil,
+                                                            views: ["view": collectionView]));
+  }
+
+  func updateTitle() {
+    if toggle == .edit {
```
Comment (romoore) [inline components/Cards/examples/EditReorderCollectionWithImagesViewController.swift]: Should this be a `switch` statement for an `enum`?

File: components/Cards/examples/EditReorderCollectionWithImagesViewController.swift [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+
+  func collectionView(_ collectionView: UICollectionView,
+                      moveItemAt sourceIndexPath: IndexPath,
+                      to destinationIndexPath: IndexPath) {
+  }
+
+  func collectionView(_ collectionView: UICollectionView,
+                      layout collectionViewLayout: UICollectionViewLayout,
+                      sizeForItemAt indexPath: IndexPath) -> CGSize {
+    let cardSize = (collectionView.bounds.size.width / 3) - 12
```
Comment (romoore) [inline components/Cards/examples/EditReorderCollectionWithImagesViewController.swift]: I'm probably missing something here, but why minus 12? If there are 3 columns and the spacing between items is 8, shouldn't it be minus 16?

File: components/Cards/examples/resources/CardsAssets.xcassets/amsterdam-zeeburg.imageset/Contents.json [hunk 19]
```diff
@@ -0,0 +1,22 @@
+{
+  "images" : [
+    {
+      "idiom" : "universal",
+      "scale" : "1x"
```
Comment (romoore) [inline components/Cards/examples/resources/CardsAssets.xcassets/amsterdam-zeeburg.imageset/Contents.json]: Missing a filename for 1x?

**Review by romoore — DISMISSED** [review]
Aside from the image licensing issue, the 1x files seem to be missing. I believe we still support iOS 8 and iPad 2 (non-retina) devices.

File: components/Cards/examples/EditReorderCollectionWithImagesViewController.swift (hunk 18 above)
Comment (galiak11) [reply components/Cards/examples/EditReorderCollectionWithImagesViewController.swift]: Not sure what's the meaning of 12. 16 didn't work either.
Anyway, this made more sense to me:
    ``` (collectionView.bounds.size.width - 8 * 4) / 3```
      - First subtract interim spaces & insets, than divide what's left to 3:

Comment (romoore) [reply components/Cards/examples/EditReorderCollectionWithImagesViewController.swift]: Sorry, so is the calculation in the code incorrect?

File: components/Cards/examples/EditReorderCollectionWithImagesViewController.swift (hunk 16 above)
Comment (galiak11) [reply components/Cards/examples/EditReorderCollectionWithImagesViewController.swift]: I believe this demo doesn't support long-touch / reordering for iOS 8 apps.

**Comment by galiak11** [comment]
Images are now approved under the same license as the source code.
Moving cards demo back to catalog now that all issues are resolved.

**Review by codeman7 — APPROVED** [review]
LGTM

**Review by yarneo — APPROVED** [review]

File: components/Cards/examples/EditReorderCollectionViewController.swift [hunk 20]
```diff
@@ -143,15 +169,10 @@ class EditReorderCollectionViewController: UIViewController,
...
 
-  func collectionView(_ collectionView: UICollectionView,
-                      moveItemAt sourceIndexPath: IndexPath,
-                      to destinationIndexPath: IndexPath) {
-  }
-
   func collectionView(_ collectionView: UICollectionView,
                       layout collectionViewLayout: UICollectionViewLayout,
                       sizeForItemAt indexPath: IndexPath) -> CGSize {
-    let cardSize = (collectionView.bounds.size.width / 3) - 12
```
Comment (romoore) [inline components/Cards/examples/EditReorderCollectionViewController.swift]: Is this some kind of bug fix? These numbers aren't equivalent and I'm still unclear why they are what they are.

File: components/Cards/examples/EditReorderCollectionViewController.swift [hunk 21]
```diff
@@ -16,23 +16,30 @@
...
   var colorScheme = MDCSemanticColorScheme()
-  let cardScheme = MDCCardScheme();
+  var typographyScheme = MDCTypographyScheme()
+  let cardScheme = MDCCardScheme()
+
+  let images = [
+    (image: "amsterdam-zeeburg",          title: "Zeeburg"),
+    (image: "venice-st-marks-square",     title: "St. Mark's Square"),
+    (image: "venice-grand-canal",         title: "Grand Canal"),
+    (image: "austin-u-texas-pond",        title: "Austin U")
```
Comment (romoore) [inline components/Cards/examples/EditReorderCollectionViewController.swift]: Should we make use of all of the images we're adding or add fewer images? I see the following are added but unused:

* amsterdam-kadoelen
* amsterdam-tuindorpoostzaan
* austin-u-texas-bronze

We're also not using "sample-image" here. Can we reuse it in this example?

File: components/Cards/src/ColorThemer/MDCCardsColorThemer.m [hunk 22]
```diff
@@ -28,6 +28,8 @@ + (void)applySemanticColorScheme:(nonnull id<MDCColorScheming>)colorScheme
 + (void)applySemanticColorScheme:(nonnull id<MDCColorScheming>)colorScheme
                       toCardCell:(nonnull MDCCardCollectionCell *)cardCell {
   cardCell.backgroundColor = colorScheme.surfaceColor;
+  cardCell.inkView.inkColor = [colorScheme.primaryColor colorWithAlphaComponent:0.08f];
+  cardCell.tintColor = colorScheme.primaryColor;
```
Comment (romoore) [inline components/Cards/src/ColorThemer/MDCCardsColorThemer.m]: Please move this to a separate PR.

File: scripts/generate_readme [hunk 23]
```diff
@@ -27,12 +27,11 @@ if [ "$#" -ne 1 ]; then
   exit 1
 fi
 
-JAZZY_VERSION=`jazzy --version`
-if [[ $? != 0 ]]; then
-  echo "Cannot find jazzy. jazzy is required to generate API links. To install try:"
-  echo "[sudo] gem install jazzy"
+command -v jazzy >/dev/null 2>&1 || {
```
Comment (romoore) [inline scripts/generate_readme]: Please move this to a separate PR.

File: components/Cards/examples/EditReorderCollectionViewController.swift [hunk 24]
```diff
@@ -116,22 +128,36 @@ class EditReorderCollectionViewController: UIViewController,
...
-    MDCCardThemer.applyScheme(cardScheme, toCardCell: cell)
-    cell.isSelectable = (toggle == .edit)
+    let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "Cell", for: indexPath)
+    guard let cardCell = cell as? CardEditReorderCollectionCell else { return cell }
+
+    MDCCardThemer.applyScheme(cardScheme, toCardCell: cardCell)
+
+    let title = dataSource[indexPath.item].title
+    let imageName = dataSource[indexPath.item].image
+    cardCell.configure(title: title, imageName: imageName)
```
Comment (romoore) [inline components/Cards/examples/EditReorderCollectionViewController.swift]: I believe you need to set the cardCell's `typographyScheme` before configuring it. Otherwise it might use an incorrect scheme.

File: components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+      post_iOS11Constraints()
+    } else {
+      pre_iOS11Constraints()
+    }
+  }
+
+  @available(iOS 11, *)
+  private func post_iOS11Constraints() {
+    #if swift(>=3.2)
+      let guide = self.safeAreaLayoutGuide
```
Comment (romoore) [inline components/Cards/examples/supplemental/CardEditReorderCollectionCell.swift]: When would a cell worry about its safe area? Normally I would expect the Collection View Controller to reposition the collection view's contents within the safe area and the cells would just render themselves regardless.  Can this be removed and just use the `pre_ios11Constraints()` instead?

File: components/Cards/README.md [hunk 26]
```diff
@@ -269,18 +269,19 @@ recommendations:
...
 easier to distinguish between elements that belong to different cards by aggregating all the 
-card's information to a single sentence.  This can be done by setting an appropriate 
+card's information so the card is read as a single sentence.  
+This can be done by setting an appropriate 
 [`accessibilityLabel`](https://developer.apple.com/documentation/uikit/uiaccessibilityelement/1619577-accessibilitylabel) 
 for the card. Additionally, set the card's 
-[`isAccessiblityElement`](https://developer.apple.com/documentation/objectivec/nsobject/1615141-isaccessibilityelement) 
-to on. Cards are a container element and setting isAccessibiltElement for a container turns 
-off individually selecting its subelements.
+[`isAccessibilityElement`](https://developer.apple.com/documentation/objectivec/nsobject/1615141-isaccessibilityelement)
```
Comment (romoore) [inline components/Cards/README.md]: Please move these changes to a separate PR.

**Review by romoore — DISMISSED** [review]

**Comment by romoore** [comment]
Based on internal decisions, it looks like the images are OK to be licensed under the Apache 2.0 license.

File: components/Cards/src/ColorThemer/MDCCardsColorThemer.m (hunk 22 above)
Comment (galiak11) [reply components/Cards/src/ColorThemer/MDCCardsColorThemer.m]: I believe this change is related to this PR. Is there anything wrong or incorrect with these?

File: components/Cards/src/ColorThemer/MDCCardsColorThemer.m [hunk 27]
```diff
@@ -1,4 +1,4 @@
-/*
+ /*
```
Comment (romoore) [inline components/Cards/src/ColorThemer/MDCCardsColorThemer.m]: Please revert this file.

**Review by romoore — APPROVED** [review]
